### PR TITLE
Add document metadata to documents table

### PR DESCRIPTION
**Pull request by lowellnava**
Connects #2547.

Add received_at, document_type, and (case) file_number fields to documents table.

Some remaining thoughts on this approach:
* We add the case file_number in the external service that fetches the documents using the vbms_id from the appeal that calls the fetch_documents_for() function. I wonder if there may be better places (a single place?) to attach the case file_number to the document.
* We don't (yet!) test that these new values make it into the database. This is mostly due to my lack of knowledge on how to test against the database (it seems like the objects we test against are not actually saved to the database?). Not testing that the values actually make it into the database isn't the worst thing in the world because the values will be used for metrics and not application-related purposes. I should probably just add tests though...

### CODE REVIEW

File: spec/models/document_spec.rb [hunk 1]
```diff
@@ -158,32 +159,30 @@
...
   context ".from_vbms_document" do
-    subject { Document.from_vbms_document(vbms_document) }
+    subject { Document.from_vbms_document(vbms_document, case_file_number) }
 
     context "when has alt doc types" do
       let(:vbms_document) do
         OpenStruct.new(
           vbms_document_id: "1",
           doc_type: "179",
-          received_at: "TEST",
```
Comment (tejans24) [inline spec/models/document_spec.rb]: Would it be helpful to make sure received_at, document_type and file_number are being persisted in the database as part of the Document model?

File: app/models/document.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@ class Document < ActiveRecord::Base
   has_many :document_views
   has_many :documents_tags
   has_many :tags, through: :documents_tags
+  before_save { self.document_type = type }
```
Comment (tejans24) [inline app/models/document.rb]: Do we still need type, if we have document_type? It seems a bit redundant, unless they have different use cases? Can you verify this?

File: db/migrate/20171026213220_add_doc_type_to_documents.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddDocTypeToDocuments < ActiveRecord::Migration
+  def change
+    add_column :documents, :document_type, :string
```
[the remark this answers is not in the document]
Comment (lowellnava) [reply db/migrate/20171026213220_add_doc_type_to_documents.rb]: I guess the main reason I wrote this as three different migrations was so that I could test each migration independently. If you think it'd be best to squash these all into a single migration before merging into master then I'd be happy to do that instead.

File: app/models/document.rb (hunk 2 above)
Comment (lowellnava) [reply app/models/document.rb]: Rails reserves the database column name "type" for single table inheritance (http://api.rubyonrails.org/classes/ActiveRecord/Inheritance.html). We could make each Document type into their own subclasses but that seemed to me a little more trouble than it is worth if all that we are using the type db field for is metric collection. However, if we foresee adding more logic based on the different Document types in the future then I'd absolutely support taking advantage of single table inheritance.

Comment (lowellnava) [reply app/models/document.rb]: To answer your question directly, document_type and type refer to the same data. We are forced to call the Document's type attribute "document_type" when storing that value in the database because of the Rails' constraints caused by single table inheritance.

Comment (tejans24) [reply app/models/document.rb]: Sorry I meant to say, why not just use document_type and remove type all together? That can be a separate issue if needed. Since we are storing the same information in two different variables.

File: db/migrate/20171026213220_add_doc_type_to_documents.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (tejans24) [reply db/migrate/20171026213220_add_doc_type_to_documents.rb]: Yea, I personally think that would be helpful to have one over multiple migration files in this scenario.

File: app/models/document.rb (hunk 2 above)
Comment (lowellnava) [reply app/models/document.rb]: Just learned that we can actually change the name of the single table inheritance column for Documents (https://apidock.com/rails/ActiveRecord/ModelSchema/ClassMethods/inheritance_column). I'm leaning towards doing that actually. Any thoughts against such an approach?

Comment (tejans24) [reply app/models/document.rb]: I personally don't see any issues with that. https://stackoverflow.com/questions/7134559/rails-use-type-column-without-sti/35232742 and a few other stackoverflow posts seem to suggest that as well. I'm fine with this or just changing all other 'type' usage to document_type, as long as we don't have two different variables for the same purpose.

File: db/migrate/20171030210701_add_received_at_to_documents.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class AddReceivedAtToDocuments < ActiveRecord::Migration
+  def change
+    add_column :documents, :received_at, :date
+    add_column :documents, :,, :string
+    add_column :documents, :document_type, :string
+    add_column :documents, :,, :string
```
Comment (tejans24) [inline db/migrate/20171030210701_add_received_at_to_documents.rb]: do we need these empty string add_colums?

File: db/migrate/20171030210701_add_received_at_to_documents.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class AddReceivedAtToDocuments < ActiveRecord::Migration
```
Comment (tejans24) [inline db/migrate/20171030210701_add_received_at_to_documents.rb]: We should probably change the name of the class and the file as well to `AddMetaDataToDocuments` or something similar.

File: db/migrate/20171030210701_add_received_at_to_documents.rb (hunk 4 above)
Comment (lowellnava) [reply db/migrate/20171030210701_add_received_at_to_documents.rb]: Nope! This looks like an artifact of me doing something wrong in the migration

File: app/models/document.rb [hunk 6]
```diff
@@ -4,6 +4,8 @@ class Document < ActiveRecord::Base
   has_many :documents_tags
   has_many :tags, through: :documents_tags
 
+  self.inheritance_column = "currently_no_inheritance"
```
Comment (tejans24) [inline app/models/document.rb]: I think this can just be nil rather than some arbitrary string but it’s not a big deal if we don’t change it.

File: spec/models/document_spec.rb [hunk 7]
```diff
@@ -158,32 +172,35 @@
...
   context ".from_vbms_document" do
-    subject { Document.from_vbms_document(vbms_document) }
+    let(:document) { Document.from_vbms_document(vbms_document, case_file_number) }
+    subject { document }
 
     context "when has alt doc types" do
       let(:vbms_document) do
         OpenStruct.new(
-          vbms_document_id: "1",
+          document_id: "1",
```
Comment (tejans24) [inline spec/models/document_spec.rb]: Why was this changed to ‘document_id’?

Comment (lowellnava) [reply spec/models/document_spec.rb]: The vbms_document variable gets passed to Document.from_vbms_document() which expects vbms_document to have a document_id and does not do anything with the vbms_document_id. I changed it here (and not in every test) because we need a document_id in order to insert into the database (which we do not do in the other tests). I'm guessing the (vbms_)?document_id incongruity may have resulted from a refactoring somewhere along the line and the extra attribute (vbms_document_id) did not negatively affect the test nor did the missing document_id attribute.

Comment (tejans24) [reply spec/models/document_spec.rb]: 👍🏼Cool, that makes sense
